### DISCUSSION TITLE
IOC scope validation design philosophy

### DISCUSSION THREAD

**Discussion by Cristiano7Zhang** [dotnet/runtime#71677] [General]
It is well known that **Scoped services aren't directly or indirectly injected into singletons**, but there is no limitation for **Transient services** as dependency of **Singleton services**. Anyone can provide more details about design philosophy for the feature?

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-extensions-dependencyinjection
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

It is well known that **Scoped services aren't directly or indirectly injected into singletons**, but there is no limitation for **Transient services** as dependency of **Singleton services**. Anyone can provide more details about design philosophy for the feature?

<table>
  <tr>
    <th align="left">Author:</th>
    <td>Cristiano7Zhang</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-Extensions-DependencyInjection`, `design-proposal`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by eerhardt** [comment]
See https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection#service-lifetimes.

In general, if you have a dependency from a Singleton service to any other service (no matter its lifetime), that service is now going to be in the "root/singleton" scope. This is fine for transient services because if someone else requests the same transient service, they will get a new instance and transient services are generally stateless.

However, scoped services are meant to hold state for the current scope (or connection). If a singleton service depends on a scoped service, that scoped service is now holding that state for the lifetime of the app. This is typically not intended.
